matt-1108-tag: custom tagging as rake task

nmaekawa opened the pull request:
- add new rake task `admin:cluster:tag` that tags instances,
  their volumes, rds instance, s3 buckets, and vpc
- include tagging of vpc, rds, and s3 in `admin:cluster:init`
- include tagging of instances in `stack:instances:start`

nmaekawa committed:
matt-1108-tag: custom tagging as rake task

- add new rake task `admin:cluster:tag` that tags instances,
  their volumes, rds instance, s3 buckets, and vpc
- include tagging of vpc, rds, and s3 in `admin:cluster:init`
- include tagging of instances in `stack:instances:start`

karendolan commented:
Noami is there a policy page or such to describe why this tagging is useful for used for?

nmaekawa commented:
besides james requesting it for billing purposes, there's this:
https://tools.dce.harvard.edu/confluence/display/desupport/AWS+Tagging

james agreed that `OU` and `Project` were the mandatory ones.

lbjay commented:
Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[lib/cluster/config_creation_session.rb, line 33 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdATycukzx2p76zmrv:-KRdATycukzx2p76zmrw:b-9hbos1) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/config_creation_session.rb#L33)):_

> ``` Ruby
>       #  @project_tag = project_tag
>       #end
>       @project_tag = project_tag
> ```

Is there a reason you decided to apply the default "MH" later in `config_creator.rb` instead of here?

---

_[lib/cluster/instances.rb, line 94 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdDteJRaDd56j5k8na:-KRdDteK5hlwmSPPVcPf:b-glkrei) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/instances.rb#L94)):_

> ``` Ruby
> 
>     def self.create_custom_tags
>       custom_tags = stack_custom_json[:aws_custom_tags] || []
> ```

I don't think you need the `[]` default here; `stack_custom_json` already returns an empty hash if it can't find the key

---

_[lib/cluster/instances.rb, line 101 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdBqz8kKqYAPzPeMXi:-KRdBqz8kKqYAPzPeMXj:b-xgtz19) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/instances.rb#L101)):_

> ``` Ruby
>       stack = Stack.find_existing
>       resource_ids = []
>       resp = opsworks_client.describe_volumes({stack_id: stack.stack_id})
> ```

It would be more consistent to move this to a `find_volumes` method

---

_[lib/cluster/rds.rb, line 96 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdDblzyuskMLLiA4zC:-KRdDblzyuskMLLiA4zD:b8l4y45) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/rds.rb#L96)):_

> ``` Ruby
> 
>     def self.create_custom_tags
>       custom_tags = stack_custom_json[:aws_custom_tags] || []
> ```

same comment re `[]`

---

_[lib/cluster/s3_bucket.rb, line 36 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdE6YNHUTfvAI72ilC:-KRdE6YNHUTfvAI72ilD:b-d85ckx) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/s3_bucket.rb#L36)):_

> ``` Ruby
>               value: stack_config[:name]
>             }
>           ].concat(stack_custom_json[:aws_custom_tags] || [])
> ```

same. i'll stop marking these.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105)_

<!-- Sent from Reviewable.io -->

lbjay commented:
OK, I realized my thinking about the `[]` default thing was wrong. I'm guessing if you got back an empty hash from `stack_custom_json` the array `.concat` method isn't going to like it. 

Instead I think you should create a `stack_custom_tags` method in `configuration_helpers.rb` and have it return the empty array by default.

nmaekawa committed:
matt-1108-tag: suggestions for pull request #105

- Stack.find_volume_ids method
- stack_custom_tags in configuration_helpers.rb that returns []

nmaekawa commented:
Reviewed 15 of 15 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[lib/cluster/config_creation_session.rb, line 33 at r1](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105#-KRdATycukzx2p76zmrv:-KRiD4gDsNgjOF-lz-U4:bc1daxm) ([raw file](https://github.com/harvard-dce/mh-opsworks/blob/a4d851b4bc230152e283903bcdbb9f168af84369/lib/cluster/config_creation_session.rb#L33)):_
<details><summary>

<i>Previously, lbjay (Jay Luker) wrote…</i></summary>



> Is there a reason you decided to apply the default "MH" later in `config_creator.rb` instead of here?
> </details>
> it has to do with templates and tests; when running tests this isn't run and i don't know how to call this in the tests so i set the default in the template, so it will always have some value in there (from the template)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/harvard-dce/mh-opsworks/105)_

<!-- Sent from Reviewable.io -->

lbjay commented:
:+1:
